src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -179,26 +186,77 @@ private void InitializeZeroTagPointIfNotInitialized()
         private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int length)
         {
             int aggregatorIndex;
-            string[] seqKey = null;
+            string[] sortedTagKeys = null;
+            object[] sortedTagValues = null;
+            ConcurrentDictionary<object[], int> value2metrics;
 
-            // GetOrAdd by TagKeys at 1st Level of 2-level dictionary structure.
-            // Get back a Dictionary of [ Values x Metrics[] ].
-            if (!this.keyValue2MetricAggs.TryGetValue(tagKeys, out var value2metrics))
+            // We only need to sort if there is more than one Tag Key.
+            if (!this.tagKeyCombinations.TryGetValue(tagKeys, out sortedTagKeys))
             {
                 // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
-                seqKey = new string[length];
+                var seqKey = new string[length];
                 tagKeys.CopyTo(seqKey, 0);
 
+                var seqValue = new object[length];
+                tagValues.CopyTo(seqValue, 0);
+
+                if (length > 1)
+                {
+                    // Create a new array for the sorted Tag keys.
+                    sortedTagKeys = new string[length];
+                    tagKeys.CopyTo(sortedTagKeys, 0);
+
+                    // Create a new array for the sorted Tag values.
+                    sortedTagValues = new object[length];
+                    tagValues.CopyTo(sortedTagValues, 0);
+
+                    Array.Sort(sortedTagKeys, sortedTagValues);
+                }
+                else
+                {
+                    sortedTagKeys = seqKey;
+                    sortedTagValues = seqValue;
+                }
+
+                this.tagValueCombinations.TryAdd(seqValue, sortedTagValues);
+                this.tagKeyCombinations.TryAdd(seqKey, sortedTagKeys);
+            }
+
+            // GetOrAdd by the sorted Tag keys at 1st Level of 2-level dictionary structure.
+            // Get back a Dictionary of [ Values x Metrics[] ].
+            if (!this.keyValue2MetricAggs.TryGetValue(sortedTagKeys, out value2metrics))
+            {
                 value2metrics = new ConcurrentDictionary<object[], int>(ObjectArrayComparer);
-                if (!this.keyValue2MetricAggs.TryAdd(seqKey, value2metrics))
+                if (!this.keyValue2MetricAggs.TryAdd(sortedTagKeys, value2metrics))
                 {
-                    this.keyValue2MetricAggs.TryGetValue(seqKey, out value2metrics);
+                    this.keyValue2MetricAggs.TryGetValue(sortedTagKeys, out value2metrics);
                 }
             }
 
+            if (!this.tagValueCombinations.TryGetValue(tagValues, out sortedTagValues))
```

Comment:
These new dictionaries may grow quite large. Could we do our check of `maxMetricPoints` earlier to avoid allowing the dictionaries to grow unbounded?

Reply:
That check won't help as we are storing different possible combinations of the tag keys. So if we have `maxMetricPoints` as `1` we should still allow for multiple entries as we have to account for different combinations. There is no way to know what the maximum possible combinations would be as it depends on the number of keys provided.

Reply:
Ah right I wasn't thinking about this correctly, so then maybe do the sort as you're doing, but defer `this.tagValueCombinations.TryAdd(seqValue, sortedTagValues);` until you know you've gotten a metric point.